File: backend/src/main/java/ru/airport/model/GateAssignment.java
package ru.airport.model;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;

/**
 * Назначение гейта на рейс с временным интервалом.
 *
 * Ключевая таблица для задачи 4 — проверки занятости гейта:
 * перед назначением сервис проверяет пересечение интервалов
 * [assigned_from, assigned_to] для данного gate_id.
 *
 * Если диспетчер меняет гейт — старая запись остаётся как история,
 * создаётся новая. Это позволяет видеть всю историю назначений.
 *
 * Также является источником данных для задачи 7 —
 * визуализации интервалов между рейсами на гейтах (timeline).
 */
@Entity
@Table(name = "gate_assignment")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@ToString(exclude = {"flight", "gate"})
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class GateAssignment {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "assignment_id")
    @EqualsAndHashCode.Include
    private Integer assignmentId;

    /**
     * Начало занятости гейта этим рейсом.
     * Обычно совпадает с scheduled_departure или чуть раньше (посадка).
     */
    @Column(name = "assigned_from", nullable = false)
    private LocalDateTime assignedFrom;

    /**
     * Конец занятости гейта этим рейсом.
     * Обычно совпадает с scheduled_departure (или фактическим вылетом).
     */
    @Column(name = "assigned_to", nullable = false)
    private LocalDateTime assignedTo;

    // ───── Связи ──────────────────────────────────────────────────────────

    /**
     * Рейс, для которого назначен гейт.
     * NOT NULL. ON DELETE CASCADE — удалён рейс → удалены все его назначения.
     */
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "flight_id", nullable = false)
    private Flight flight;

    /**
     * Гейт, назначенный на рейс.
     * NOT NULL. ON DELETE RESTRICT — нельзя удалить гейт с активными назначениями.
     */
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "gate_id", nullable = false)
    private Gate gate;
}

File: clients/mobile/mobile_personal_app/lib/models/gate_assignment.dart
class GateSummary {
  final int gateId;
  final String gateNumber;
  final String? terminal;

  GateSummary({required this.gateId, required this.gateNumber, this.terminal});

  factory GateSummary.fromJson(Map<String, dynamic> json) {
    return GateSummary(
      gateId: json['gateId'] as int? ?? 0,
      gateNumber: json['gateNumber'] as String? ?? '',
      terminal: json['terminal'] as String?,
    );
  }
}

class GateAssignment {
  final int assignmentId;
  final String? assignedFrom;
  final String? assignedTo;
  final GateSummary? gate;

  GateAssignment({
    required this.assignmentId,
    this.assignedFrom,
    this.assignedTo,
    this.gate,
  });

  factory GateAssignment.fromJson(Map<String, dynamic> json) {
    final gateJson = json['gate'] as Map<String, dynamic>?;
    return GateAssignment(
      assignmentId: json['assignmentId'] as int? ?? 0,
      assignedFrom: json['assignedFrom']?.toString(),
      assignedTo: json['assignedTo']?.toString(),
      gate: gateJson != null ? GateSummary.fromJson(gateJson) : null,
    );
  }
}
